File: 2021/src/bin/day1.rs
use std::error::Error;
use std::fs::File;
use std::io::{BufRead as _, BufReader};
use std::path::Path;

fn load_depths<P: AsRef<Path>>(input: P) -> Result<Vec<u32>, Box<dyn Error>> {
    let mut depths = Vec::new();
    for line in BufReader::new(File::open(input)?).lines() {
        depths.push(line?.parse()?);
    }
    Ok(depths)
}

pub mod part1 {
    // O(N) time, O(1) space
    pub fn solve<I>(depths: I) -> usize
    where
        I: IntoIterator<Item = u32>,
    {
        depths
            .into_iter()
            .scan(None, |old: &mut Option<u32>, new| {
                Some(old.replace(new).filter(|&old| new > old).is_some())
            })
            .filter(|&b| b)
            .count()
    }

    #[cfg(test)]
    mod tests {
        use super::super::load_depths;
        use super::solve;

        #[test]
        fn test_solve() {
            let depths = load_depths("tests/day1/sample").unwrap();
            assert_eq!(7, solve(depths.iter().cloned()));
        }
    }
}

pub mod part2 {
    use super::part1;

    const WINDOW_SIZE: usize = 3;

    // O(N) time, O(1) space; but requires input taking O(N) space
    pub fn solve(depths: &[u32]) -> usize {
        part1::solve(depths.windows(WINDOW_SIZE).map(|win| win.iter().sum()))
    }

    #[cfg(test)]
    mod tests {
        use super::super::load_depths;
        use super::solve;

        #[test]
        fn test_solve() {
            assert_eq!(5, solve(&load_depths("tests/day1/sample").unwrap()));
        }
    }
}

fn main() {
    let input = "tests/day1/input";
    let depths = load_depths(input).unwrap_or_else(|err| {
        eprintln!("error: {}: {}", input, err);
        std::process::exit(3);
    });
    println!("{}", part1::solve(depths.iter().cloned()));
    println!("{}", part2::solve(&depths));
}

File: 2021/src/bin/day1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func loadDepths(fileName string) ([]int, error) {
	r, err := os.Open(fileName)
	if err != nil {
		return nil, err
	}

	defer r.Close()

	var ds []int
	s := bufio.NewScanner(r)
	s.Split(bufio.ScanWords)
	for s.Scan() {
		d, err := strconv.Atoi(s.Text())
		if err != nil {
			return ds, err
		}

		ds = append(ds, d)
	}

	return ds, s.Err()
}

func solvePart1(depths []int) uint {
	var count uint
	for i := 1; i < len(depths); i++ {
		if depths[i] > depths[i-1] {
			count++
		}
	}

	return count
}

func solvePart2(depths []int) uint {
	var ds []int
	for i := 2; i < len(depths); i++ {
		ds = append(ds, depths[i-2]+depths[i-1]+depths[i])
	}

	return solvePart1(ds)
}

func main() {
	depths, err := loadDepths("tests/day1/input")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	println(solvePart1(depths))
	println(solvePart2(depths))
}
